src/transactions/OfferExchange.cpp

```diff
@@ -1391,4 +1454,129 @@ convertWithOffers(
+
+struct ExchangedQuantities
+{
+    int64_t sheepSend{0};
+    int64_t wheatReceived{0};
+};
+
+static bool
+shouldConvertWithOffers(std::optional<ExchangedQuantities> const& poolExchange,
+                        ExchangedQuantities const& bookExchange)
+{
+    if (poolExchange)
+    {
+        // pE.sS * bE.wR > pE.wR * bE.sS is equivalent in arbitrary precision to
+        // bE.wR / bE.sS > pe.wR / pE.sS, but the first form avoids rounding in
+        // finite precision
+        //
+        // We want to use the liquidity pool unless the order book provides a
+        // strictly better price
+        return bigMultiply(poolExchange->sheepSend,
+                           bookExchange.wheatReceived) >
+               bigMultiply(poolExchange->wheatReceived, bookExchange.sheepSend);
+    }
+
+    // If we can't exchange with the pool then either
+    // - the exchange with the book succeeded so we should use that
+    // - the exchange with the book failed so we should return that error code
+    //   for compatibility with old behavior
+    return true;
+}
+
+ConvertResult
+convertWithOffersAndPools(
+    AbstractLedgerTxn& ltxOuter, Asset const& sheep, int64_t maxSheepSend,
+    int64_t& sheepSend, Asset const& wheat, int64_t maxWheatReceive,
+    int64_t& wheatReceived, RoundingType round,
+    std::function<OfferFilterResult(LedgerTxnEntry const&)> filter,
+    std::vector<ClaimAtom>& offerTrail, int64_t maxOffersToCross)
+{
+    ZoneScoped;
+
+    // If offerTrail is not empty at the start, then the limit maxOffersToCross
+    // will not be imposed correctly.
+    releaseAssertOrThrow(offerTrail.empty());
+
+    sheepSend = 0;
+    wheatReceived = 0;
+
+    // Compute the exchange from the liquidity pool but don't actually do the
+    // exchange
+    std::optional<ExchangedQuantities> poolExchange;
+    {
+        ExchangedQuantities res;
+        if (exchangeWithPool(ltxOuter, sheep, maxSheepSend, res.sheepSend,
+                             wheat, maxWheatReceive, res.wheatReceived, round))
+        {
+            poolExchange = std::make_optional<ExchangedQuantities>(res);
+        }
+    }
+
+    // Maybe use the order book
+    {
+        LedgerTxn ltxConvertWithOffers(ltxOuter);
+
+        ExchangedQuantities bookExchange;
+        auto convertRes = convertWithOffers(
+            ltxConvertWithOffers, sheep, maxSheepSend, bookExchange.sheepSend,
+            wheat, maxWheatReceive, bookExchange.wheatReceived, round, filter,
+            offerTrail, maxOffersToCross);
+
+        if (shouldConvertWithOffers(poolExchange, bookExchange))
+        {
+            sheepSend = bookExchange.sheepSend;
+            wheatReceived = bookExchange.wheatReceived;
+            ltxConvertWithOffers.commit();
+            return convertRes;
+        }
+    }
+
+    // Clear offerTrail because it was initially empty and we didn't actually
+    // cross the order book
+    offerTrail.clear();
+    if (static_cast<uint64_t>(maxOffersToCross) <= 1)
```

Comment:
this is incorrect: it should be `if (maxOffersToCross == 0)` as `maxOffersToCross` represent the maximum size allowed for `offerTrail` *inclusive* (to mimic `convertWithOffers`)